scenario/scripts/snapshot.py

```diff
@@ -0,0 +1,752 @@
+import json
+import logging
+import os
+import re
+import tempfile
+from dataclasses import asdict
+from enum import Enum
+from itertools import chain
+from pathlib import Path
+from subprocess import CalledProcessError, check_output, run
+from textwrap import dedent
+from typing import Any, BinaryIO, Dict, Iterable, List, Optional, TextIO, Tuple, Union
+
+import ops.pebble
+import typer
+import yaml
+
+from scenario.state import (
+    Address,
+    BindAddress,
+    Container,
+    Model,
+    Mount,
+    Network,
+    Relation,
+    State,
+    Status,
+)
+
+logger = logging.getLogger("snapshot")
+
+JUJU_RELATION_KEYS = frozenset({"egress-subnets", "ingress-address", "private-address"})
+JUJU_CONFIG_KEYS = frozenset({})
+SNAPSHOT_TEMPDIR_ROOT = (Path(os.getcwd()).parent / "snapshot_storage").absolute()
+
+
+class SnapshotError(RuntimeError):
+    pass
+
+
+class InvalidTarget(SnapshotError):
+    pass
+
+
+class InvalidModel(SnapshotError):
+    pass
+
+
+class Target(str):
+    def __init__(self, unit_name: str):
+        super().__init__()
+        app_name, _, unit_id = unit_name.rpartition("/")
+        if not app_name or not unit_id:
+            raise InvalidTarget(f"invalid unit name: {unit_name!r}")
+        self.unit_name = unit_name
+        self.app_name = app_name
+        self.unit_id = int(unit_id)
+        self.normalized = f"{app_name}-{unit_id}"
+
+
+def _try_format(string):
+    try:
+        import black
+
+        try:
+            return black.format_str(string, mode=black.Mode())
+        except black.parsing.InvalidInput as e:
+            logger.error(f"error parsing {string}: {e}")
+            return string
+    except ModuleNotFoundError:
+        logger.warning("install black for formatting")
+        return string
+
+
+def format_state(state):
+    return _try_format(repr(state))
+
+
+def format_test_case(state, charm_type_name=None, event_name=None):
+    ct = charm_type_name or "CHARM_TYPE  # TODO: replace with charm type name"
+    en = event_name or "EVENT_NAME,  # TODO: replace with event name"
+    return _try_format(
+        dedent(
+            f"""
+from scenario.state import *
+from charm import {ct}
+
+def test_case():
+    state = {state}
+    out = state.trigger(
+        {en}
+        {ct}
+        )
+
+"""
+        )
+    )
+
+
+def _juju_run(cmd, model=None) -> Dict[str, Any]:
+    _model = f" -m {model}" if model else ""
+    raw = run(
+        f"""juju {cmd}{_model} --format json""".split(), capture_output=True
+    ).stdout.decode("utf-8")
+    return json.loads(raw)
+
+
+def _juju_ssh(target: Target, cmd, model: Optional[str] = None) -> str:
+    _model = f" -m {model}" if model else ""
+    command = f"""juju ssh{_model} {target.unit_name} {cmd}"""
+    raw = run(command.split(), capture_output=True).stdout.decode("utf-8")
+    return raw
```

Comment:
Couple notes:

Generally triple quotes should only be used for docstrings or in-memory "documents" (Jinja templates, Python files, etc.); they should not be used for regular strings.

It is encouraged to use the [`shlex`](https://docs.python.org/3/library/shlex.html#shlex.split) module in the Python standard library for parsing strings intended to be run as shell commands rather than `str.split()`. `shlex.split()` splits the string using shell-like syntax. `shlex` can ignore comments and also use POSIX syntax.

Also, what is the type of `cmd`? See the other arguments have an annotation, but `cmd` does not.

Lastly, you could shorten this method by quite a bit (i.e. no one-user variables):

```suggestion
def _juju_ssh(target: Target, cmd: str, model: Optional[str] = None) -> str:
    return run(
        shlex.split(
            f"juju ssh{(f' -m {model}' if model else '')} {target.unit_name} {cmd}"
        ),
        capture_output=True,
    ).stdout.decode("utf-8")
```

Reply:
"you could shorten this method quite a bit" and goes on to add 2 lines :D

but yeah, it does look nicer. I always forget shlex is there. however, I hate nested fstrings, I think it hinders readability.
So +1 for shlex, -1 for removing _model.

Reply:
Sounds good. You should use shlex definitely, but I won't be sad about you not using nested f-strings. The nested f-strings are my preferences if you are just evaluating simple conditionals, but to each there own :)

_Black did me dirty by adding those two extra lines_

Reply:
Yep, please not nested f-strings! :-)

I don't think we should be using `shlex.split` *or* `str.split` here -- `subprocess.run` already takes a list of arguments and handles shlex-ing correctly, so use that. It avoids the `ssh{model}` spacing messiness too.

```python
def _juju_ssh(target: Target, cmd, model: Optional[str] = None) -> str:
    model_args = ["-m", model] if model else []
    process = run(["juju", "ssh", *model_args, target.unit_name, cmd],
                  capture_output=True, text=True)
    return process.stdout
```

You *could* use `shlex.split` on just the `cmd` part, but there's actually no need; it's fine to pass a shell command line to `juju ssh` as a single argument. But change `cmd` to `*shlex.split(cmd)` if you prefer.

Also node the use of `text=True`, avoiding the need for `.decode("utf-8")`.